### PR TITLE
Characteristics read/write/monitor

### PR DESCRIPTION
**Pull request by bartoszwilk**

### CODE REVIEW

File: android/src/main/java/com/polidea/flutter_ble_lib/CharacteristicsResponse.java [hunk 1]
```diff
@@ -3,20 +3,28 @@
...
+import java.util.UUID;
+
 public class CharacteristicsResponse {
     private final Characteristic[] characteristics;
-    private final Service service;
+    private int serviceId;
+    private UUID serviceUuid;
 
-    public CharacteristicsResponse(Characteristic[] characteristics, Service service) {
+    public CharacteristicsResponse(Characteristic[] characteristics, int serviceId, UUID serviceUuid) {
```
[the remark this answers is not in the document]
Comment (bartoszwilk) [reply android/src/main/java/com/polidea/flutter_ble_lib/CharacteristicsResponse.java]: We just only need these 2 informations. Already discussed.

File: android/src/main/java/com/polidea/flutter_ble_lib/delegate/CharacteristicsDelegate.java [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+    }
+
+    private void writeCharacteristicForIdentifier(int characteristicIdentifier,
+                                                  byte[] bytesToWrite,
+                                                  boolean withResponse,
+                                                  String transactionId,
+                                                  final MethodChannel.Result result) {
+        bleAdapter.writeCharacteristic(
+                characteristicIdentifier,
+                Base64Converter.encode(bytesToWrite),
```
[the remark this answers is not in the document]
Comment (bartoszwilk) [reply android/src/main/java/com/polidea/flutter_ble_lib/delegate/CharacteristicsDelegate.java]: Created.

File: android/src/main/java/com/polidea/flutter_ble_lib/delegate/DiscoveryDelegate.java [hunk 3]
```diff
@@ -106,87 +107,59 @@ public void onError(BleError error) {
...
-                });
+        try {
+            Characteristic[] characteristics = adapter.getCharacteristicsForService(serviceId);
+            result.success(new CharacteristicJsonConverter().toJson(characteristics));
+        } catch (BleError error) {
+            error.printStackTrace();
+            result.error(String.valueOf(error.errorCode.code), error.reason, bleErrorJsonConverter.toJson(error));
+        } catch (JSONException e) {
+            e.printStackTrace();
+            result.error(e.toString(), e.getMessage(), null);
```
[the remark this answers is not in the document]
Comment (bartoszwilk) [reply android/src/main/java/com/polidea/flutter_ble_lib/delegate/DiscoveryDelegate.java]: Created.

File: lib/internal/flutter_ble_lib_mixins/characteristics_mixin.dart [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    String transactionId,
+  }) =>
+      _methodChannel.invokeMethod(
+        MethodName.readCharacteristicForService,
+        <String, dynamic>{
+          ArgumentName.serviceId: serviceIdentifier,
+          ArgumentName.characteristicUuid: characteristicUUID,
+          ArgumentName.transactionId: transactionId
+        },
+      ).then((rawJsonValue) {
```
[the remark this answers is not in the document]
Comment (bartoszwilk) [reply lib/internal/flutter_ble_lib_mixins/characteristics_mixin.dart]: What do you mean by that? It is decoded from json one line below

File: android/src/main/java/com/polidea/flutter_ble_lib/SingleCharacteristicResponse.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import com.polidea.multiplatformbleadapter.Characteristic;
+
+import java.util.UUID;
+
+public class SingleCharacteristicResponse {
+    private final Characteristic characteristic;
+    private int serviceId;
+    private UUID serviceUuid;
+
+    public SingleCharacteristicResponse(Characteristic characteristics, int serviceId, UUID serviceUuid) {
```
Comment (dariuszseweryn) [inline android/src/main/java/com/polidea/flutter_ble_lib/SingleCharacteristicResponse.java]:
```suggestion
    public SingleCharacteristicResponse(Characteristic characteristic, int serviceId, UUID serviceUuid) {
```

File: android/src/main/java/com/polidea/flutter_ble_lib/SingleCharacteristicResponse.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        this.characteristic = characteristics;
```
Comment (dariuszseweryn) [inline android/src/main/java/com/polidea/flutter_ble_lib/SingleCharacteristicResponse.java]:
```suggestion
        this.characteristic = characteristic;
```

File: example/lib/sensor_tag_test_scenario.dart [hunk 7]
```diff
@@ -52,12 +53,24 @@ class TestScenario {
...
       }
     });
-
     log("CONNECTING to ${peripheral.name}, ${peripheral.identifier}...");
     await peripheral.connect();
     log("CONNECTED to ${peripheral.name}, ${peripheral.identifier}!");
     deviceConnectionAttempted = false;
 
+    monitoringStreamSubscription?.cancel();
+    monitoringStreamSubscription = peripheral
```
Comment (dariuszseweryn) [inline example/lib/sensor_tag_test_scenario.dart]: Maybe I do not know Dart enough but it looks weird. Is `Peripheral` a `StreamSubscription`?

Comment (bartoszwilk) [reply example/lib/sensor_tag_test_scenario.dart]: No. peripheral.monitorCharacteristic returns a stream and Stream.listen() returns a subscription

File: lib/internal/managers_for_classes.dart [hunk 8]
```diff
@@ -27,7 +29,7 @@ abstract class ManagerForPeripheral {
     String transactionId,
   );
 
-  Future<Characteristic> writeCharacteristicForDevice(
+  Future<CharacteristicWithValue> writeCharacteristicForDevice(
```
Comment (mikolak) [inline lib/internal/managers_for_classes.dart]: Should be `Future<Characteristic>`

File: lib/internal/managers_for_classes.dart [hunk 9]
```diff
@@ -53,7 +55,7 @@ abstract class ManagerForService {
     String transactionId,
   );
 
-  Future<Characteristic> writeCharacteristicForService(
+  Future<CharacteristicWithValue> writeCharacteristicForService(
```
Comment (mikolak) [inline lib/internal/managers_for_classes.dart]: Should be `Future<Characteristic>`

File: lib/internal/managers_for_classes.dart [hunk 10]
```diff
@@ -77,7 +79,7 @@ abstract class ManagerForCharacteristic {
     String transactionId,
   );
 
-  Future<Characteristic> writeCharacteristicForIdentifier(
+  Future<CharacteristicWithValue> writeCharacteristicForIdentifier(
```
Comment (mikolak) [inline lib/internal/managers_for_classes.dart]: Should be `Future<void>`

File: lib/internal/internal_ble_manager.dart [hunk 11]
```diff
@@ -136,7 +138,7 @@ class InternalBleManager
       );
 
   @override
-  Future<Characteristic> writeCharacteristicForIdentifier(Peripheral peripheral,
+  Future<CharacteristicWithValue> writeCharacteristicForIdentifier(Peripheral peripheral,
```
Comment (mikolak) [inline lib/internal/internal_ble_manager.dart]: Should be `Future<void>`

File: lib/internal/internal_ble_manager.dart [hunk 12]
```diff
@@ -148,7 +150,7 @@ class InternalBleManager
       );
 
   @override
-  Future<Characteristic> writeCharacteristicForDevice(
+  Future<CharacteristicWithValue> writeCharacteristicForDevice(
```
Comment (mikolak) [inline lib/internal/internal_ble_manager.dart]: Should be `Future<Characteristic>`

File: lib/internal/internal_ble_manager.dart [hunk 13]
```diff
@@ -165,7 +167,7 @@ class InternalBleManager
       );
 
   @override
-  Future<Characteristic> writeCharacteristicForService(
+  Future<CharacteristicWithValue> writeCharacteristicForService(
```
Comment (mikolak) [inline lib/internal/internal_ble_manager.dart]: Should be `Future<Characteristic>`

File: lib/internal/internal_ble_manager.dart [hunk 14]
```diff
@@ -158,53 +158,53 @@ class InternalBleManager
...
       _bleLib.monitorCharacteristicForService(
         peripheral,
         serviceIdentifier,
         characteristicUUID,
-        transactionId: transactionId,
+        transactionId,
       );
 
   @override
   Stream<CharacteristicWithValue> monitorCharacteristicForIdentifier(
```
Comment (mikolak) [inline lib/internal/internal_ble_manager.dart]: Should be `Stream<UInt8List>`

File: lib/service.dart [hunk 15]
```diff
@@ -29,17 +29,16 @@ class Service {
     bool withResponse, {
     String transactionId,
   }) =>
-      _manager.writeCharacteristicForService(
-          peripheral, _id, characteristicUUID, bytes, withResponse,
-          transactionId: transactionId);
+      _manager.writeCharacteristicForService(peripheral, _id,
+          characteristicUUID, bytes, withResponse, transactionId);
 
   Future<Characteristic> readCharacteristic(String characteristicUUID,
```
Comment (mikolak) [inline lib/service.dart]: Should be `Future<CharacteristicWithValue>`

File: android/src/main/java/com/polidea/flutter_ble_lib/delegate/CharacteristicsDelegate.java [hunk 16]
```diff
@@ -323,7 +319,13 @@ private void monitorCharacteristicForIdentifier(int characteristicIdentifier,
                 transactionId, new OnEventCallback<Characteristic>() {
                     @Override
                     public void onEvent(Characteristic data) {
-                        characteristicsMonitorStreamHandler.onCharacteristicsUpdate(data);
+                        try {
+                            characteristicsMonitorStreamHandler.onCharacteristicsUpdate(
+                                    createCharacteristicResponse(data)
+                            );
+                        } catch (JSONException e) {
+                            e.printStackTrace();
```
Comment (mikolak) [inline android/src/main/java/com/polidea/flutter_ble_lib/delegate/CharacteristicsDelegate.java]: Should be passed to stream

File: android/src/main/java/com/polidea/flutter_ble_lib/delegate/CharacteristicsDelegate.java [hunk 17]
```diff
@@ -347,7 +349,13 @@ private void monitorCharacteristicForDevice(String deviceIdentifier,
                 new OnEventCallback<Characteristic>() {
                     @Override
                     public void onEvent(Characteristic data) {
-                        characteristicsMonitorStreamHandler.onCharacteristicsUpdate(data);
+                        try {
+                            characteristicsMonitorStreamHandler.onCharacteristicsUpdate(
+                                    createCharacteristicResponse(data)
+                            );
+                        } catch (JSONException e) {
+                            e.printStackTrace();
```
Comment (mikolak) [inline android/src/main/java/com/polidea/flutter_ble_lib/delegate/CharacteristicsDelegate.java]: Should be passed to stream

File: example/lib/sensor_tag_test_scenario.dart (hunk 7 above)
Comment (mikolak) [reply example/lib/sensor_tag_test_scenario.dart]: > Maybe I do not know Dart enough but it looks weird. Is `Peripheral` a `StreamSubscription`?

This is a chainging of calls. `peripheral.monitorCharacteristic().listen()` - last call returns a stream subscription.

File: android/src/main/java/com/polidea/flutter_ble_lib/delegate/CharacteristicsDelegate.java [hunk 18]
```diff
@@ -0,0 +1,432 @@
...
+                                                 final MethodChannel.Result result) {
+        bleAdapter.monitorCharacteristicForService(
+                serviceIdentifier,
+                characteristicUuid,
+                transactionId,
+                new OnEventCallback<Characteristic>() {
+                    @Override
+                    public void onEvent(Characteristic data) {
+                        try {
+                            characteristicsMonitorStreamHandler.onCharacteristicsUpdate(
```
Comment (mikolak) [inline android/src/main/java/com/polidea/flutter_ble_lib/delegate/CharacteristicsDelegate.java]: Is it thread safe?

File: android/src/main/java/com/polidea/flutter_ble_lib/delegate/DiscoveryDelegate.java [hunk 19]
```diff
@@ -104,87 +106,57 @@ public void onError(BleError error) {
...
-                    }
-                }, new OnErrorCallback() {
-                    @Override
-                    public void onError(BleError error) {
-                        result.error(String.valueOf(error.errorCode.code), error.reason, bleErrorJsonConverter.toJson(error));
-                    }
-                });
+        try {
+            Service[] services = adapter.getServicesForDevice(deviceId);
+            result.success(serviceJsonConverter.toJson(services));
```
Comment (mikolak) [inline android/src/main/java/com/polidea/flutter_ble_lib/delegate/DiscoveryDelegate.java]: Can services be null/empty and not trigger an exception? If yes, what should happen?

File: android/src/main/java/com/polidea/flutter_ble_lib/delegate/DiscoveryDelegate.java [hunk 20]
```diff
@@ -104,87 +106,57 @@ public void onError(BleError error) {
...
-                    }
-                },
-                new OnErrorCallback() {
-                    @Override
-                    public void onError(BleError error) {
-                        result.error(String.valueOf(error.errorCode.code), error.reason, bleErrorJsonConverter.toJson(error));
-                    }
-                });
+        try {
+            Characteristic[] characteristics = adapter.getCharacteristicsForService(serviceId);
```
Comment (mikolak) [inline android/src/main/java/com/polidea/flutter_ble_lib/delegate/DiscoveryDelegate.java]: Can characteristics be empty/null? If yes, what should happen? What happens if serviceId is unknown?

File: example/lib/device_details/device_details_bloc.dart [hunk 21]
```diff
@@ -112,6 +114,39 @@ class DeviceDetailsBloc {
         })
         .then((characteristics) => characteristics.forEach((characteristic) =>
             log("Found characteristic \n ${characteristic.uuid}")))
+        .then((_) {
+          log("Turn off temperature update");
+          return peripheral.writeCharacteristic(
```
Comment (mikolak) [inline example/lib/device_details/device_details_bloc.dart]: You're demoing only one approach, ie. reading/writing from peripheral level using known UUIDs. You should also test against those operation from Service and Characteristic level.

File: example/lib/sensor_tag_test_scenario.dart [hunk 22]
```diff
@@ -79,6 +92,39 @@ class TestScenario {
         })
         .then((characteristics) => characteristics.forEach((characteristic) =>
             log("Found characteristic \n ${characteristic.uuid}")))
+        .then((_) {
+          log("Turn off temperature update");
+          return peripheral.writeCharacteristic(
```
Comment (mikolak) [inline example/lib/sensor_tag_test_scenario.dart]: Same thing as in device_details_bloc - only one approach is tested.

File: lib/internal/constants.dart [hunk 23]
```diff
@@ -46,7 +61,12 @@ abstract class ArgumentName {
 
   static const String serviceUuid = "serviceUuid";
   static const String serviceId = "serviceId";
```
Comment (mikolak) [inline lib/internal/constants.dart]: serviceIdentifier?

File: android/src/main/java/com/polidea/flutter_ble_lib/delegate/DiscoveryDelegate.java (hunk 19 above)
Comment (bartoszwilk) [reply android/src/main/java/com/polidea/flutter_ble_lib/delegate/DiscoveryDelegate.java]: Services can be empty and cannot be null. After the discussion with @Cierpliwy we decided that this is an expected behavior.

File: lib/internal/constants.dart (hunk 23 above)
Comment (dariuszseweryn) [reply lib/internal/constants.dart]: Isn't it even some kind of `internalUniqueServiceIdentifier`?

File: android/src/main/java/com/polidea/flutter_ble_lib/delegate/DiscoveryDelegate.java (hunk 20 above)
Comment (bartoszwilk) [reply android/src/main/java/com/polidea/flutter_ble_lib/delegate/DiscoveryDelegate.java]: Ad1. Characteristics can be empty, but cannot be null
Ad.2 If serviceId is null then BleError will be thrown.

File: lib/internal/constants.dart (hunk 23 above)
Comment (bartoszwilk) [reply lib/internal/constants.dart]: We can consider a better naming for that but for now I will just leave serviceIdentifier for the sake of consistency with similar variables within the lib and MBA

File: example/lib/util/pair.dart [hunk 24]
```diff
@@ -0,0 +1,9 @@
+class Pair<T, S> {
```
Comment (mikolak) [inline example/lib/util/pair.dart]: We don't need this.

File: example/lib/sensor_tag_test_scenario.dart [hunk 25]
```diff
@@ -79,6 +95,118 @@ class TestScenario {
...
+        .then((service) async {
+          log("Turn on temperature update");
+          Characteristic configCharacteristic =
+              await service.writeCharacteristic(
+                  SensorTagTemperatureUuids.temperatureConfigCharacteristic,
+                  Uint8List.fromList([1]),
+                  false);
+          return Pair(service, configCharacteristic);
+        })
+        .then((serviceAndConfigCharacteristic) =>
```
Comment (mikolak) [inline example/lib/sensor_tag_test_scenario.dart]: This flow can be simplified by using more of async await. We don't need to chain all of those calls, but the example should read nicely and I don't think it does now. Perhaps some operations should be extracted to separate functions?

File: example/lib/device_details/device_details_bloc.dart [hunk 26]
```diff
@@ -112,6 +115,116 @@ class DeviceDetailsBloc {
...
+        .then((service) =>
+            Future.delayed(Duration(seconds: 1)).then((_) => service))
+        .then((service) async {
+          CharacteristicWithValue dataCharacteristic =
+              await service.readCharacteristic(
+                  SensorTagTemperatureUuids.temperatureDataCharacteristic);
+          return Pair(service, dataCharacteristic);
+        })
+        .then((serviceAndCharacteristic) {
+          log("Temperature value ${serviceAndCharacteristic.second.value}");
```
[the remark this answers is not in the document]
Comment (bartoszwilk) [reply example/lib/device_details/device_details_bloc.dart]: I created issue to refactor this test case.

File: example/lib/util/pair.dart (hunk 24 above)
Comment (bartoszwilk) [reply example/lib/util/pair.dart]: I created issue to refactor this test case.

File: example/lib/sensor_tag_test_scenario.dart (hunk 25 above)
Comment (bartoszwilk) [reply example/lib/sensor_tag_test_scenario.dart]: I created issue to refactor this test case.

**Pull request merged**